python/tank/templatekey.py

```diff
@@ -339,41 +355,217 @@ def __init__(self,
                                         abstract=abstract,
                                         length=length)
 
+        if self._subset_format and not self._subset_str:
+            raise TankError("%s: Cannot specify subset_format parameter without a subset parameter." % self)
+
+
     @property
     def filter_by(self):
         """
         Name of filter type to limit values for string.
-
-        returns: 'alphanumeric', 'alpha', None or a regex string.
+        ``alphanumeric``, ``alpha``, ``None`` or a regex string.
         """
         return self._filter_by
 
+    @property
+    def subset(self):
+        """
+        Returns a regular expression describing how values should be transformed
+        when they are being injected into template paths and strings.
+
+        The format for a subset is a regular expression containing regex groups,
+        for example::
+
+            # grabs capital letters of the two first words
+            user_initials:
+                type: str
+                subset: '([A-Z])[a-z]* ([A-Z])[a-z]*'
+
+            # extracts the first three characters
+            first_three_characters:
+                type: str
+                subset: '(.{3}).*'
+
+            # in code, the above expressions would compress the following input:
+
+            some_template.apply_fields(
+                {"user_initials": "John Smith",
+                "first_three_characters": "John Smith"}
+            )
+
+            # into "JS" for the {user_initials} key and "Joh" for the {first_three_characters} key
+
+        If the subset expression contains more than one ``(regex group)`` to extract, the groups
+        will be concatenated together in the order they are found. If you want greater control
+        over this, see :meth:`subset_format`.
+        """
+        return self._subset_str
+
+    @property
+    def subset_format(self):
+        """
+        Returns the ``subset_format`` string for the given template key. This string is used in conjunction with the
+        :meth:`subset` parameter and allows for the formatting of the values that are being extracted::
+
+            # grabs capital letters of the two first words
+            user_initials_backwards:
+                type: str
+                subset: '([A-Z])[a-z]* ([A-Z])[a-z]*'
+                subset_format: '{1} {0}'
+
+            # in code, the above expression would compress the following input:
+
+            some_template.apply_fields({"user_initials": "John Smith"})
+
+            # into "SJ" for the user_initials_backwards key.
```

Comment:
Quick note that this is the place where the result for the example is wrong (missing space)